### ISSUE TITLE
[PT-D]Track PT-D fixes for release blocker to upgrade to PT 3.11

### ISSUE THREAD

**Issue by wz337** [pytorch/pytorch#92319]
**Overview**
Python 3.11 was [released in Oct 2022](https://www.python.org/downloads/release/python-3110/). It is the first release to include work from the “Faster CPython” workgroup. They advertise speedups between 10 and 60% depending on the workload without any change to the user code.

The issue for adding support in PyTorch [#86566](https://github.com/pytorch/pytorch/issues/86566) is already getting to the top of the list of issues when sorted by 👍 and 🚀reactions. We are also getting almost one issue a day at this point about people asking for conda support for 1.13 and/or because some features are broken in the current pip package we provide.
In the short term, with PyTorch 2.0 being a performance focused release, native python speedup can only be beneficial for the user experience.

Some of the main features of interest added here are:
Frozen imports / Static code objects (to speed up startup)
Cheaper, lazy Python frames
Inlined Python function calls
Specializing Adaptive Interpreter (a very basic jit)
More concise exception representation

Of course adding a new version of Python is a significant amount of work on the Infrastructure side (both for core and extended) and for this particular release, the low level changes in CPython lead to issues for PyTorch, especially for components like TorchScipt and Dynamo.

**The goal of this document is to have a full list of what needs to be done for 3.11 support to be on par with all the other python versions.
The goal is also to align on what we consider release blocker for 2.0 (if not everything can be done by then).**

Details for all distributed related failures are as follows:

**DISTRIBUTED** [#92301](https://github.com/pytorch/pytorch/issues/92301) 
Failures related to dataclasses
distributed/test_distributed_spawn for example
  File "/home/albandes/local/installs/python3.11/release/install/lib/python3.11/dataclasses.py", line 815, in _get_field
    raise ValueError(f'mutable default {type(f.default)} for field '
ValueError: mutable default <class 'torch.distributed._shard.sharded_tensor.metadata.TensorProperties'> for field tensor_properties is not allowed: use default_factory
Most likely due to https://docs.python.org/3/whatsnew/3.11.html#dataclasses
One error test_launch_user_script
  File "/home/albandes/local/pytorch/3.11_release_source/torch/distributed/elastic/multiprocessing/api.py", line 162, in to_map
    map[i] = val_or_map.get(i, Std.NONE)
AttributeError: 'NoneType' object has no attribute 'get'
One error test_from_value
  File "/home/albandes/local/pytorch/3.11_release_source/test/distributed/elastic/multiprocessing/api_test.py", line 94, in test_from_value
    self.assertEqual(Std.NONE, Std.from_str("0"))

cc @ezyang @gchanan @zou3519 @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu

**Comment by rohan-varma** [comment]
cc @albanD

**Comment by rohan-varma** [comment]
Also related: https://github.com/pytorch/pytorch/issues/92301

Could we consolidate those here and close the above issue? cc oncall @fduwjj

**Comment by awgu** [comment]
This PR https://github.com/pytorch/pytorch/pull/91795 looks good to me to fix the first error, but I will let @wanchaol or @fduwjj to approve.

**Comment by fduwjj** [comment]
OK, upon further investigation, below is my findings for the last two test errors:
1. It is PT 3.11 specific.
2. It is related to this Int Flag:
https://github.com/pytorch/pytorch/blob/master/torch/distributed/elastic/multiprocessing/api.py#L100-L104

So the root cause is that under PT 3.11, the first value 0 and last value 0 gets missing during the for loop iteration here: https://github.com/pytorch/pytorch/blob/master/torch/distributed/elastic/multiprocessing/api.py#L122-L124

That's why we see two test fails and it belongs to elastic folks, I don't know if changing it to IntEnum will cause any other problems but IntEnum works in this case. cc r2p oncall here: @kurman 

Below is simple code for repro and issue illustration:

In PT3.9:

```
Python 3.9.12 (main, Jun  1 2022, 11:38:51)
[GCC 7.5.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from enum import IntFlag
>>> class Color(IntFlag):
...     RED = 0
...     BLUE = 1
...     GREY = 2
...     YELLOW = 3
...
>>> for i in Color:
...     print(i)
...
Color.RED
Color.BLUE
Color.GREY
Color.YELLOW
```

While in PT 3.11:
```
Python 3.11.0 (main, Jan 16 2023, 14:19:54) [GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>>
>>>
>>>
>>> from enum import IntFlag
>>> class Color(IntFlag):
...     RED = 0
...     BLUE = 1
...     GREY = 2
...     YELLOW = 3
...
>>> for i in Color:
...     print(i)
...
1
2
```

That's why we see test failures. I can do the change if we are ok with changing it from `IntFlag` to `IntEnum` Or maybe this is a bug of PyThon which we need to report an issue?

**Comment by jayaddison** [comment]
@fduwjj I think you've nearly tracked it down there: as far as I can tell, this is _not_ a bug, but a subtle behaviour relating to iteration and aliases:

https://docs.python.org/3/howto/enum.html#flag-and-intflag-minutia

(both `RED` and `YELLOW` in your example are aliases, I think)

**Comment by fduwjj** [comment]
> @fduwjj I think you've nearly tracked it down there: as far as I can tell, this is _not_ a bug in Python 3.11, but a subtle behaviour relating to iteration and aliases:
> 
> https://docs.python.org/3/howto/enum.html#flag-and-intflag-minutia
> 
> (both `RED` and `YELLOW` in your example are aliases, I think)

Got it. Thanks for explaining. I am wondering what's the recommended way to fix it because the original code does want to use the enum also as Flag.

**Comment by jayaddison** [comment]
You're welcome.

Although I think there are opportunities for more refactoring and improvement, we're only attempting a migration for Python3.11 (not any other improvements/features), so it probably makes sense to be cautious about changing any logic.

Looking at the existing code:

https://github.com/pytorch/pytorch/blob/550f98332bfd296bf10b50ae43f5677e1f04f4bd/torch/distributed/elastic/multiprocessing/api.py#L120-L125

It does a few things:

- Allocates a variable that holds an instance (in fact, a [singleton](https://docs.python.org/3.7/library/enum.html#enum-members-aka-instances)) `s` of type `Std` at any given time
- Scans the domain of possible `Std` values
- Finds the singleton `s` that is equal to `v` (it uses `v == s`, although `v is s` should also evaluate to `True` when that does)
- Returns that singleton

Retaining the domain-checking behaviour and singleton-instance matching behaviour seem particularly important.

Also, zero values have some behaviour in `enum.IntFlag` that we should be careful about.  A zero bitmask _always_ exists within other bitmasks (as mentioned in the [minutia](https://docs.python.org/3.11/howto/enum.html#flag-and-intflag-minutia)).

Before implementation, let's check across a range of values with a domain that includes zero using a Python script, and then run that script on both Python3.10 and Python3.11.

```py
# test.py
from enum import IntFlag
class Test92319(IntFlag):
    ZERO = 0
    ONE = 1
    TWO = 2
    THREE = 3
 
values = (Test92319(v) for v in [0, 1, 2, 3])
assert all(value in Test92319 for value in values)  # we can use the Python 'in' operator to check membership of the values in the enum
assert Test92319(4) not in Test92319  # this value should _not_ be found within the members of the enum
```

# Conclusion
If the `test.py` script is good enough evidence for you -- and running it yourself and/or adding more test coverage wouldn't hurt to confirm that -- then I'd recommend:

```py
        def to_std(v):
            s = Std(int(v))
            if s in Std:
                return s
            # return None -> should NEVER reach here since we regex check input
```

In summary: instead of iterating through the members of `Std` and then returning one that matches the input:

- Use the input to construct an `Std` value (which will be a singleton if it is a member of the enum)
- Perform an (efficient) lookup to check whether it's a known `Std` value
  - If so, return it as the result.

**Comment by jayaddison** [comment]
> and then run that script on both Python3.10 and Python3.11

Ek, one more thing: probably worth running the `test.py` script -- and/or any other additional assurance checks for this -- on supported versions going back to Python3.7, to match the [package's Python requirements](https://github.com/pytorch/pytorch/blob/78caa7921c03cf46463a0d88c35d171b4a798cdc/setup.py#L217).

**Comment by jayaddison** [comment]
@fduwjj By the way, your comment at https://github.com/pytorch/pytorch/issues/92319#issuecomment-1396243726 (comparing what looks like sensible behaviour in Py3.9 with what looks like a weird behaviour change in Py3.11) inspired an attempt to detect the cause from `pylint` - still a work-in-progress, maybe useful in future? https://github.com/PyCQA/pylint/pull/8105

**Comment by jayaddison** [comment]
Please note: there is a pull request with a potential solution to this high-priority blocker, but please confirm that any and all questions raised in #85559 are resolved before looking at the changes included there.

**Comment by fduwjj** [comment]
@jayaddison Thanks for providing so detailed explanation.. This is indeed something I wasn't aware of. Thanks!!

**Comment by jayaddison** [comment]
> Perform an (efficient) lookup

It might be worth checking my claim about this.  I _thought_ that `IntFlag` membership tests should be trivially efficient; I'm less certain about that currently, though.

Related to this, I'm reading python/cpython#93910 at the moment (to clarify: I don't yet know for sure that there is a problem with the efficiency of the lookups in the proposed change.  but it would be worth confirming)

**Comment by jayaddison** [comment]
> > Perform an (efficient) lookup
> 
> It might be worth checking my claim about this. I _thought_ that `IntFlag` membership tests should be trivially efficient; I'm less certain about that currently, though.
> 
> Related to this, I'm reading [python/cpython#93910](https://github.com/python/cpython/issues/93910) at the moment (to clarify: I don't yet know for sure that there is a problem with the efficiency of the lookups in the proposed change. but it would be worth confirming)

After reading through that issue thread, I believe that the performance issues raised there relate to beta versions of 3.11, and that they were resolved for the stable releases of Python 3.11 (such as the ones that PyTorch will use).

So I think that all is well, and that using the membership test (`if s in Std`) approach is fine.  Again, worth someone double-checking me on that.

**Comment by jayaddison** [comment]
If someone else wants to develop a fix implementation of their own -- say, for verification or training purposes -- then some approximate guidance I'd be happy to share (and wouldn't mind seeing merged by someone else) would be:

> Build an instance of class `Std` using the integer value of `v` as the only argument.  If that instance is considered a member of the class `Std`, then return the instance as the result.

**Comment by fduwjj** [comment]
Two changes have been landed and close this for now. Feel free to reopen it.
